packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts

```diff
@@ -35,12 +35,28 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
   private _source: HierarchyDefinition;
   private _nodeIdentifierPaths: HierarchyFilteringPath[];
   private _pathsIdentifierPositions: Dictionary<HierarchyNodeIdentifier, Array<[number, number]>>;
+  /**
+   * Cache that is used for determining how many identifiers from two paths are similar.
+   * Identifiers are similar if:
+   * a) They are both of [[IModelInstanceKey]] type and:
+   *  1. They have the same id's and imodelKeys.
+   *  2. They have the same or related (derived) class.
+   * b) They are both of [[GenericNodeKey]] type and:
+   *  1. They have the same id's and sources.
+   * For example, if provided these pathIndexes:
+   * {lhsIndex: 0, rhsIndex: 1}
+   * with these _nodeIdentifierPaths:
+   * [[a, b, c, d, e], [a, b, c, e, f]]
+   * This dictionary would return: 3, since both paths start with a, b, c identifiers (which are similar)
+   */
+  private _pathsSimilarityLengthCache: Dictionary<{ lhsIndex: number; rhsIndex: number }, number>;
 
   public constructor(props: FilteringHierarchyDefinitionProps) {
     this._imodelAccess = props.imodelAccess;
     this._source = props.source;
     this._nodeIdentifierPaths = props.nodeIdentifierPaths;
     this._pathsIdentifierPositions = this.createPathsIdentifierPositions(props.nodeIdentifierPaths);
+    this._pathsSimilarityLengthCache = this.createPathsSimilarityLengthCache();
```

Comment:
At this point we have class 3 members + a ton of code for handling those paths. Maybe it would be worth moving all of that into a separate class responsible just for that?

Reply:
Created a separate class for handling positions